### ISSUE TITLE
Feature Request:

### ISSUE THREAD

**Issue by The1iAmNeo** [matplotlib/mplfinance#572]
i want to create a footprint chart like this one where i am able to make an open high low close but also show how much trading volume was at each y value of the candle

here is an example of what i am talking about

![image](https://user-images.githubusercontent.com/95321908/204052918-3fad026f-4d1b-4393-acb0-dea401dda0bd.png)

its very easy to get the data ... getting and understanding how to organize the data on our end is very very easy ... you can see what a csv file would look like here ... https://public.bybit.com/trading/BTCUSD/

lets say i wanted to create a 5 min candle ... for the info i would just add up all the sell orders at each price point of the candle and do the same for the buy orders ... then i have a dataframe within the candle that tells me the sell column and buy column ... and you can get the open high low close even from that data ... the open is the first trade at the start of the candle then the high is the highest point ... low is the lowest point and the close is the last candle ...

this info helps us visualize at what price point in the candle has the most trade volume ... footprint charts are used in high level trading ... I will be making a ton of videos on this subject if we can get one of these open source plotting platforms to help create something like this

all we would need is send dataframes inside a master dataframe ... so you have the overall dataframe that houses all of the candles and inside each candle is a dataframe of all of the sell and buy data at each price of the candle

ultimately would be nice for it to look like this ... but for now the top one would be ultra sexy
![image](https://user-images.githubusercontent.com/95321908/204053003-0e5ba28c-0b9d-4d8f-940a-6c369cd9d588.png)

**Comment by DanielGoldfarb** [comment]
@The1iAmNeo 
Very interesting!  I will take a closer look in a week or so (busy with another project for the next week or so).  At that time I will look into how best to implement this, and where it should fall in priorities.  How are your python coding skills?   Perhaps you can help with some of the coding?
Thanks.   All the best.  --Daniel

**Comment by The1iAmNeo** [comment]
@DanielGoldfarb hey thanks for taking an interest ... it would really help out the algo trading community so much.

my coding skills aren't the greatest but would love to chat and see what i could help with ... only been coding for like 2 months so not sure what i could do lol ... but i am willing to try because having this would be a life saver

**Comment by diliprk** [comment]
> 

I am interested in developing this feature as well, I work as a Data Engineer at `nFerence.ai` . However, I don't fully understand the underlying mechanisms, logic, and math behind this foot prints chart so if I could get some guidance on that front I can collaborate on the coding. BTW, is it possible to make the charts interactive like in plotly or bokeh?

**Comment by The1iAmNeo** [comment]
@DanielGoldfarb hey ... wanted to check back in and see if you were still interested in doing this?

**Comment by zaheeru** [comment]
Would be interested to see this up and running, please let know how I can help , not a good coder in python though

**Comment by andrewrgarcia** [comment]
@QuantFreedom1022 I'd like to help to code this Candlestick Footprint Chart method, but I need some clarification. Are we treating the bid-traded and ask-traded volumes as those indicated by the Buy and Sell labels in the 'side' column of the [provided csv files](https://dxfeed.com/order-historical-market-data/sample-data/)? Also, is there a specific number of market orders that must be plotted for each footprint or a rule associated with such?

**Comment by AudiKid** [comment]
this chart is actually a great idea to see if market makers are net long or net short . depending how many orders have been bought at each candle stick and the adding of buy orders at a certain price this indicates a high successful rate of the market hitting those number . market makers work like a grocery store they buy cheap to sale for more ,,, or if they buy high at a short they sell to cover their short . 

this chart will give u a visualization of market and with more coding maybe see what prices are getting bought or sold the most and hypothetically they will hit those number .

**Comment by QuantFreedom1022** [comment]
@DanielGoldfarb ... wanted to check back in with you on this ... would like to start working on this soon

@andrewrgarcia yeah i am down to work together on this ... i already have the code and the csv files needed ... just need the plotting to do its job

**Comment by DanielGoldfarb** [comment]
I am very involved right now working on a project for [Jupytercon](https://www.jupytercon.com/).  I am happy for you two to begin moving forward with this, and I can put *some* time into it.  I would like to see some details regarding design and what you are planning, so as to fit nicely into the big picture and long term plans for mplfinance.  You can post here, or if you prefer, email me at dgoldfarb.github@gmail.com

The first thing I'd like know is what is the minimum expectation for the data (what columns are a must; which are optional).  If the data is going to look like this, https://public.bybit.com/trading/BTCUSD/, then it seems to me that it doesn't necessarily fit into the standard `mpf.plot(df,**kwargs)` function, which expects simply OHLC(V) data.  Perhaps we write a wrapper around `mpf.plot()`?  Or, if you have in mind just adding a new `type=` to `mpf.plot()`, maybe that can work as well, but I'd like to see what that might look like inside.

Looking forward to hearing from you.  All the best.  --Daniel

**Comment by andrewrgarcia** [comment]
> @DanielGoldfarb ... wanted to check back in with you on this ... would like to start working on this soon
> 
> @andrewrgarcia yeah i am down to work together on this ... i already have the code and the csv files needed ... just need the plotting to do its job

Nice. You can start by attaching the files and I'll see what I can do.

**Comment by andrewrgarcia** [comment]
> @andrewrgarcia and @DanielGoldfarb ... here is the link to my github repo for the footprints ... and also here is a quick loom video of me explaining a little of how it works https://github.com/QuantFreedom1022/Footprints https://www.loom.com/share/76ccbf0a46374256b798f2a237f41f78

@QuantFreedom1022 I forked your repository [here](https://github.com/andrewrgarcia/Footprints) to shorten and revise your footprint code. My [revised version](https://github.com/andrewrgarcia/Footprints/blob/main/footprint.py) does not use numba yet it runs faster than your former code. It also removes the sell / buy count issue you mention in your video and thus, appropriately reduces the trade count (see csv output screen captures for the [revised code](https://raw.githubusercontent.com/andrewrgarcia/Footprints/main/tests/footprintcsv.png) and [your code](https://raw.githubusercontent.com/andrewrgarcia/Footprints/main/tests/footprintscsv_quantfreedom.png)). Still, please check the code and provide me some feedback if needed before I submit a PR to merge it to your repository

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia WOW ... ummm so this code is way way way above my skill level lol ... so i feel it almost impossible to give any feedback lol ... 

but i did take your code and add it to my repo so that way we have your code and i also have my code ... but yeah lets move forward with yours seeing as how it is faster and better in all aspects.

also i would love to jump on a quick 10 or so screen share call to talk about what we are both thinking as far as moving the project forward now that we have a function that can organize the data to start to build the footprint charts

**Comment by andrewrgarcia** [comment]
> @andrewrgarcia WOW ... ummm so this code is way way way above my skill level lol ... so i feel it almost impossible to give any feedback lol ...
> 
> but i did take your code and add it to my repo so that way we have your code and i also have my code ... but yeah lets move forward with yours seeing as how it is faster and better in all aspects.
> 
> also i would love to jump on a quick 10 or so screen share call to talk about what we are both thinking as far as moving the project forward now that we have a function that can organize the data to start to build the footprint charts

@QuantFreedom1022 Hey, it's possible you're new to open source. Can you please remove the [andrew_footprint.py](https://github.com/QuantFreedom1022/Footprints/blob/main/andrew_footprint.py) file you created by copying my code and instead approve my Pull Request (PR) ? I just sent you one. 

Best.

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia ok i just deleted your file and approved your pull request ... so everything should be up to date ... again for me i would prefer a quick 10 min screen share call for next steps but its up to you ... if you would like to setup a time to talk you can let me know your email and we can set that up or we can just do it here.

if not then what are your thoughts on what to do next

**Comment by andrewrgarcia** [comment]
> @andrewrgarcia WOW ... ummm so this code is way way way above my skill level lol ... so i feel it almost impossible to give any feedback lol ... 
> 
> 
> 
> but i did take your code and add it to my repo so that way we have your code and i also have my code ... but yeah lets move forward with yours seeing as how it is faster and better in all aspects.
> 
> 
> 
> also i would love to jump on a quick 10 or so screen share call to talk about what we are both thinking as far as moving the project forward now that we have a function that can organize the data to start to build the footprint charts

I think the biggest challenge now would be on how to lump the timestamps into groups with the volume data so that footprints are made. My first thought was to use an argmax and/or argmin function, but if you have any other ideas and/or can sketch it, that would help. 

The other thing would be on the side of the new footprints.py code to make it more lean. Your initial algorithm used several global variables, and because of the lengthiness of it, the new code inherits all these variables as constructors to the Printer class i.e. the variables below the __init__() function; some of these variables do not need to be constructors as some of them may be local to the specific code section they are used on. It would be great if you could identify the ones which do not need to be so "global" as this was originally your algorithm.

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia okay well I'm going to have to take some time to read through your code and figure out exactly what it's doing because I've never done object-oriented programming before so I have to educate myself a little bit and then once I feel like I have a grasp of that I'll make a loom video explaining anything that I see and also maybe some suggestions as well on how to put the footprints together

But this would be good because I'm at a point with my other project that I have to learn an object oriented programming so this is a good simple code to practice figuring out what's going on

**Comment by andrewrgarcia** [comment]
> @andrewrgarcia okay well I'm going to have to take some time to read through your code and figure out exactly what it's doing because I've never done object-oriented programming before so I have to educate myself a little bit and then once I feel like I have a grasp of that I'll make a loom video explaining anything that I see and also maybe some suggestions as well on how to put the footprints together
> 
> But this would be good because I'm at a point with my other project that I have to learn an object oriented programming so this is a good simple code to practice figuring out what's going on

@QuantFreedom1022 That's great. The second paragraph is more of a request. I would prioritize the stuff I mentioned on the first paragraph (ideas/algos/drawn sketches for the lumping of timestamps with volumes). This can be done without learning the new code and more by looking at the csv that you generate with your former code and how to put the info together for the footprints

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia ok sounds good ... ill have a look at it as soon as I can and let u know any ideas I have

**Comment by andrewrgarcia** [comment]
> i want to create a footprint chart like this one where i am able to make an open high low close but also show how much trading volume was at each y value of the candle
> 
> 
> 
> here is an example of what i am talking about
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/95321908/204052918-3fad026f-4d1b-4393-acb0-dea401dda0bd.png)
> 
> 
> 
> its very easy to get the data ... getting and understanding how to organize the data on our end is very very easy ... you can see what a csv file would look like here ... https://public.bybit.com/trading/BTCUSD/
> 
> 
> 
> lets say i wanted to create a 5 min candle ... for the info i would just add up all the sell orders at each price point of the candle and do the same for the buy orders ... then i have a dataframe within the candle that tells me the sell column and buy column ... and you can get the open high low close even from that data ... the open is the first trade at the start of the candle then the high is the highest point ... low is the lowest point and the close is the last candle ...
> 
> 
> 
> this info helps us visualize at what price point in the candle has the most trade volume ... footprint charts are used in high level trading ... I will be making a ton of videos on this subject if we can get one of these open source plotting platforms to help create something like this
> 
> 
> 
> all we would need is send dataframes inside a master dataframe ... so you have the overall dataframe that houses all of the candles and inside each candle is a dataframe of all of the sell and buy data at each price of the candle
> 
> 
> 
> ultimately would be nice for it to look like this ... but for now the top one would be ultra sexy
> 
> ![image](https://user-images.githubusercontent.com/95321908/204053003-0e5ba28c-0b9d-4d8f-940a-6c369cd9d588.png)
> 
> 

I did this but in a more elegant way (my last PR in candle builder.py). Instead of a dataframe within a master dataframe, the code creates a "3-d" dataframe so to speak, where each open and close time is now unique and each entry for every other column is now a list of ask bid and volumes thereof. The next steps would be making the actual graphs and answering the questions I posted in this new PR.

**Comment by Philip030601** [comment]
Hi all,  my name us philip and loojs like im comming in late.  I was planning on coding some footprint logic and came accross this project.  Do u guys still need help? Im an intermediate developer at best but very resourceful and have a tendency to get things done.  Look forward to you reply/ies.  

Much thanks

**Comment by QuantFreedom1022** [comment]
> Hi all, my name us philip and loojs like im comming in late. I was planning on coding some footprint logic and came accross this project. Do u guys still need help? Im an intermediate developer at best but very resourceful and have a tendency to get things done. Look forward to you reply/ies.
> 
> Much thanks

@Philip030601 yeah man that would be great ... i need to get back into this and working with someone else on it would help me get back into this project ... could you send me your discord or telegram or email or something and then we can setup a time to chat about what is needed to be done and share some ideas on what we want it to look like

**Comment by Philip030601** [comment]
Hi quantfreedom.

Hope your doing well.  Im in pst time and available to talk after 6pm
almost every night.  Let me know of your availability and will plan on it. B

Thanks

**Comment by QuantFreedom1022** [comment]
@Philip030601 ok ... let me know a contact for you that would be good ... either telegram, email or discord

**Comment by Philip030601** [comment]
> @Philip030601 ok ... let me know a contact for you that would be good ... either telegram, email or discord

Hi quant. My email is sflush369@gmail.com.

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia not sure if you are still interested in this but i am picking the project back up ... and i now understand about classes and all the stuff i didn't know about 6 months ago ... so i can actually be useful this time ... let me know if you are still interested in trying to finish this

**Comment by QuantFreedom1022** [comment]
I found a person who has something close to what i want ... you can check out the project here https://github.com/murtazayusuf/OrderflowChart

**Comment by andrewrgarcia** [comment]
I am happy to contribute with you guys, but contingent on 1) that we all follow good open-source / version control practices so that we all get credit for our contributions and 2) we start from scratch with an object-oriented structure with concise code with little to no additional dependencies (packages). Mplfinance is a lean package and does not use numba, which in principle can make it quite heavy. In addition, I didn't find the implementation of numba helpful as I made it just as fast without it. Code cannot be clunky, it needs to be sleek

**Comment by QuantFreedom1022** [comment]
@andrewrgarcia yeah i am happy to do so

1: this will be my first time trying to work with someone over the internet with only typing and also using github for version controls and practices and all that stuff ... so if you are willing to work with someone that is 100% new to this then i am fine but i have absolutely no idea about what good version control practices are ... i don't even know rule number one

2: yes i can do this but the concise code with little to no dependencies not sure ... all i know is i want to use numpy and as little pandas as possible because pandas is so slow ... but you are smarter than me so i am ok with using what ever dependencies that you want

We already have the project open at https://github.com/QuantFreedom1022/Footprints .. but i have no idea where to start ... or where should we start talking about this idea?

if you could lead this that would be great because i had no idea how to lead a project with only typing over the internet ... i am more of a lets hope on zoom ... talk for 15 mins and then get to work

**Comment by KMey-create** [comment]
Hi there, I will need some help with this, where can I get this type of data?
I was trying Tradermade.com, but there are Bid/Ask/Mid columns

**Comment by QuantFreedom1022** [comment]
@KMey-create i get mine from binance https://www.binance.com/en/landing/data

**Comment by Philip030601** [comment]
Quant,

Happy belated thanksgiving.  I thought the data had to be time and sales
type in tick format.

Regards,

**Comment by Vimal-Seshadri-Raguraman** [comment]
@QuantFreedom1022 @andrewrgarcia 
Hello everyone, I recently stumbled upon this discussion while exploring ways to create a footprint chart in Python, specifically for stock data, not cryptocurrency. I'm eager to contribute and offer assistance in any way I can. While I wouldn't classify myself as an expert programmer, I believe I have a good understanding and can navigate my way through this. Feel free to reach out to me at rsehshadri@gmail.com for any further discussion or help.

**Comment by QuantFreedom1022** [comment]
@Vimal-Seshadri-Raguraman i don't think Andrew will be working in this but I am still interested ... pls email me and we can setup a time to talk ... quantfreedom1022@gmail.com

**Comment by jamesdidit72** [comment]
Has there been any progress with this?

**Comment by quantfreedom** [comment]
![image](https://github.com/matplotlib/mplfinance/assets/95321908/90680c11-f3ff-42d1-89a2-dd245d8f205f)
we have made our own version if you or anyone else wants to help build it out more, you can join my discord and message me there ... the link to my linktree with my discord is on my github profile

once the concept is built out more would be nice to present it to matplotlib and see if we could work together on this idea

**Comment by AudiKid** [comment]
id be interested in using this build 

    On Sunday, June 23, 2024 at 12:07:21 PM CDT, Neo | Quant Freedom ***@***.***> wrote:  
 
 


@jamesdidit72
image.png (view on web)
we have made our own version if you or anyone else wants to help build it out more, you can join my discord and message me there ... the link to my linktree with my discord is on my github profile

**Comment by premanand2008** [comment]
> id be interested in using this build  On Sunday, June 23, 2024 at 12:07:21 PM CDT, Neo | Quant Freedom ***@***.***> wrote: @jamesdidit72 image.png (view on web) we have made our own version if you or anyone else wants to help build it out more, you can join my discord and message me there ... the link to my linktree with my discord is on my github profile — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***>

@jamesdidit72 would like to join and contribute in this project. How do I join?

**Comment by Vimal-Seshadri-Raguraman** [comment]
Hey guys,

Sorry, I had to step away for a while. I haven't been able to focus/follow thru on this idea. But I am back and I am actively interested in building footprint chart and orderflow charts. Let me know if you guys are still interested. 

Sorry for not replying back @quantfreedom

**Comment by quantfreedom** [comment]
@Vimal-Seshadri-Raguraman please check my previous message to u and email me if ur still interested ... we have built out an app with way more than footprint candles that u might be interested in

**Comment by Vimal-Seshadri-Raguraman** [comment]
Yes I saw it. It is interesting. I am yet to test it and go through the code. And I will email about what I had in mind
